service/frontend/handler.go

```diff
@@ -587,11 +583,25 @@ func (wh *WorkflowHandler) GetWorkflowExecutionHistory(
 			DomainUUID: common.StringPtr(info.ID),
 			Execution:  getRequest.GetExecution(),
 		})
-		if err != nil {
-			return nil, wrapError(err)
+		if err == nil {
+			token.nextEventID = response.GetEventId()
+			token.runID = response.GetRunId()
+		} else {
+			if _, ok := err.(*gen.EntityNotExistsError); !ok || !getRequest.GetExecution().IsSetRunId() {
+				return nil, wrapError(err)
+			}
+			// It is possible that we still have the events in the table even though the mutable state is gone
+			// Get the nextEventID from visibility store if we still have it.
+			visibilityResp, err := wh.visibitiltyMgr.GetClosedWorkflowExecution(&persistence.GetClosedWorkflowExecutionRequest{
```

Comment:
I believe this call is used by both decision workers as well as by the visibility client (CLI). Do you want to make the assumption here that the caller is always interested in the ClosedWFExecution() ? For decision workers, this API call must still return an error if the WFExecution does not exist. How about passing in a special flag in the request struct, that gives you a hint that the caller is interested in closedWFExecutions as well ?